lib/util/cache/repository/impl/s3.ts

```diff
@@ -0,0 +1,72 @@
+import { Readable } from 'stream';
+import {
+  GetObjectCommand,
+  GetObjectCommandInput,
+  PutObjectCommand,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { logger } from '../../../../logger';
+import { getS3Client, parseS3Url } from '../../../s3';
+import { streamToString } from '../../../streams';
+import type { RepoCacheRecord } from '../types';
+import { RepoCacheBase } from './base';
+
+export class RepoCacheS3 extends RepoCacheBase {
+  private readonly s3Client;
+  private readonly bucket;
+
+  constructor(repository: string, url: string) {
+    super(repository);
+    this.bucket = parseS3Url(url)?.Bucket;
+    this.s3Client = getS3Client();
+  }
+
+  async read(): Promise<string | undefined> {
+    const cacheFileName = this.getCacheFileName();
+    const s3Params: GetObjectCommandInput = {
+      Bucket: this.bucket,
+      Key: cacheFileName,
+    };
+    try {
+      const { Body: res } = await this.s3Client.send(
+        new GetObjectCommand(s3Params)
+      );
+      if (res instanceof Readable) {
+        const data: string = JSON.parse(await streamToString(res));
```

Comment:
```suggestion
        const data = JSON.parse(await streamToString(res));
```

i think we should move the parsing to base load function in a separate PR

Reply:
sure, ill do it.

Reply:
https://github.com/renovatebot/renovate/pull/17346